### ISSUE TITLE
TypeError when creating security group ingress rule via CloudFormation

### ISSUE THREAD

**Issue by yasuder** [spulec/moto#3542]
- **How to reproduce the issue**
Using a CFT like this:
```
{
    "AWSTemplateFormatVersion": "2010-09-09",
    "Description": "AWS CloudFormation Template to create an EC2 instance",
    "Parameters": {
        "SubnetId": {
            "Type": "String",
            "Description": "An internal subnet ID of a subnet to deploy in",
            "AllowedPattern": "^subnet-[a-zA-Z0-9]*"
        },
        "VPCId": {
            "Type": "String",
            "Description": "The VPC ID",
            "AllowedPattern": "^vpc-[a-zA-Z0-9]*"
        },
        "EC2KeyName": {
            "Type": "String",
            "Description": "The key name assigned to the EC2 instance",
            "Default": "My-Key-Pair"
        },
        "InstanceRoleName": {
            "Type": "String",
            "Description": "The name of the IAM role used by the instance",
            "Default": "My-IAM-Role"
        }
    },
    "Resources": {
        "EC2Instance": {
            "Type": "AWS::EC2::Instance",
            "Properties": {
                "IamInstanceProfile": {
                    "Ref": "IAMInstanceProfile"
                },
                "ImageId": "ami-03657b56516ab7912",
                "InstanceType": "t2.micro",
                "KeyName": {
                    "Ref": "EC2KeyName"
                },
                "SecurityGroupIds": [
                    {
                        "Fn::GetAtt": [
                            "SecurityGroup",
                            "GroupId"
                        ]
                    }
                ],
                "SubnetId": {
                    "Ref": "SubnetId"
                },
                "Tags": [
                    {
                        "Key": "Name",
                        "Value": "My-EC2-Instance"
                    }
                ],
                "Tenancy": "default"
            }
        },
        "SecurityGroup": {
            "Type": "AWS::EC2::SecurityGroup",
            "Properties": {
                "GroupDescription": "Default VPC security group for the instance",
                "GroupName": "My-SG",
                "VpcId": {
                    "Ref": "VPCId"
                }
            }
        },
        "SSHIngressRule": {
            "Type": "AWS::EC2::SecurityGroupIngress",
            "Properties": {
                "CidrIp": "10.0.0.0/8",
                "Description": "Allow SSH traffic from 10.0.0.0/8",
                "FromPort": 22,
                "ToPort": 22,
                "GroupId": {
                    "Fn::GetAtt": [
                        "SecurityGroup",
                        "GroupId"
                    ]
                },
                "IpProtocol": "tcp"
            }
        },
        "IAMInstanceProfile": {
            "Type": "AWS::IAM::InstanceProfile",
            "Properties": {
                "InstanceProfileName": "My-Instance-Profile",
                "Roles": [
                    {
                        "Ref": "InstanceRoleName"
                    }
                ]
            }
        }
    }
}
```
If I have this snippet in a `create_stack.py` file and run this in main (where TEMPLATE is the above JSON file that I've run a `json.load` and `json.dumps` on):
```
import boto3

def create_my_stack(subnet_id, vpc_id):
    cf_client.create_stack(
        StackName=self.stack_name,
        TemplateBody=TEMPLATE,
        Parameters=[
            {
                "ParameterKey": "SubnetId",
                "ParameterValue": subnet_id
            },
            {
                "ParameterKey": "VPCId",
                "ParameterValue": vpc_id
            }
        ],
        Capabilities=['CAPABILITY_NAMED_IAM'],
        OnFailure='DELETE'
    )
```
I have no issues. However if I try to use moto and pytest to assert that I get a response back, I get the following error:
```
self = <moto.ec2.models.EC2Backend object at 0x109fe8890>, group_name_or_id = 'sg-7463a68f', ip_protocol = 'tcp'
from_port = 22, to_port = 22, ip_ranges = ['10.0.0.0/8'], source_group_names = None, source_group_ids = None
vpc_id = None

    def authorize_security_group_ingress(
        self,
        group_name_or_id,
        ip_protocol,
        from_port,
        to_port,
        ip_ranges,
        source_group_names=None,
        source_group_ids=None,
        vpc_id=None,
    ):
        group = self.get_security_group_by_name_or_id(group_name_or_id, vpc_id)
        if ip_ranges:
            if isinstance(ip_ranges, str) or (
                six.PY2 and isinstance(ip_ranges, unicode)  # noqa
            ):
                ip_ranges = [{"CidrIp": str(ip_ranges)}]
            elif not isinstance(ip_ranges, list):
                ip_ranges = [json.loads(ip_ranges)]
        if ip_ranges:
            for cidr in ip_ranges:
>               if not is_valid_cidr(cidr["CidrIp"]):
E               TypeError: string indices must be integers
```
   - **What you expected to happen**
I expected there to be a response, similar to what I get when I actually run the method outside of pytest.

   - **What actually happens**
I get the TypeError from above. It looks like it's looking for a `CidrIp` field in a [dict](https://github.com/spulec/moto/blob/master/moto/ec2/models.py#L2234), rather than the actual CIDR value itself from a list (which is what's being provided). 

   - **What version of Moto you're using** and **How you installed it**
moto==1.3.16
boto3==1.16.34
botocore==1.19.34

I am using the `mock_cloudformation` and `mock_ec2` (to create my fake VPC and subnet to deploy in) and setting them as a pytest fixture to pass into my test case.
In my `test_create_stack.py`
```
def test_create_stack_success(cloudformation, ec2):
    subnet = ec2.describe_subnets()['Subnets'][0]
    subnet_id = subnet['SubnetId']
    vpc_id = subnet['VpcId']
    assert "StackId" in create_stack(subnet_id, vpc_id)
```

In my `conftest.py`:
```
...
@pytest.fixture
def ec2():
    """
    Pytest fixture that creates the test VPC and subnet in the
    fake moto AWS account

    Yields a fake boto3 ec2 client
    """
    with mock_ec2():
        ec2 = boto3.client("ec2", region_name="us-west-2")
        vpc = ec2.create_vpc(
            CidrBlock="172.16.0.0/16",
            AmazonProvidedIpv6CidrBlock=False,
            InstanceTenancy="default"
        )
        vpc_id = vpc['Vpc']['VpcId']
        ec2.create_subnet(
            CidrBlock="172.16.0.0/17",
            VpcId=vpc_id,

        )
        yield ec2
...
```

**Comment by bblommers** [comment]
Thanks for raising this @yasuder - marking it as a bug.

Just out of curiosity, are you able to reproduce this with moto 1.3.15.dev1021 or earlier? It looks like this is a regression introduced in #3196

**Comment by yasuder** [comment]
I just tried with moto==1.3.15 and that still threw the same error. I was able to get it to work with moto==1.3.14, but that threw a few different errors.
```
Tried to parse AWS::EC2::SecurityGroupEgress but it's not supported by moto's CloudFormation implementation
```
and 
```
Tried to parse AWS::IAM::Policy but it's not supported by moto's CloudFormation implementation
```

Looks like these features are just unavailable with moto with this version though, so I'll have to test the call with an altered CloudFormation template.

**Comment by bblommers** [comment]
The TypeError should now be fixed as of moto >= 1.3.16.dev199
